src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,652 @@
+    /// less liquidity assets reserved for the ARM's withdrawal queue.
+    /// This does not exclude any accrued performance fees.
+    function _availableAssets() internal view returns (uint256) {
+        // Get the assets in the ARM and external withdrawal queue
+        uint256 assets = token0.balanceOf(address(this)) + token1.balanceOf(address(this)) + _externalWithdrawQueue();
+
+        // Load the queue metadata from storage into memory
+        uint256 queuedMem = withdrawsQueued;
+        uint256 claimedMem = withdrawsClaimed;
+
+        // If the ARM becomes insolvent enough that the available assets in the ARM and external withdrawal queue
+        // is less than the outstanding withdrawals and accrued fees.
+        if (assets + claimedMem < queuedMem) {
+            return 0;
+        }
+
+        // Need to remove the liquidity assets that have been reserved for the withdrawal queue
+        return assets + claimedMem - queuedMem;
+    }
+
+    /// @dev Hook for calculating the amount of assets in an external withdrawal queue like Lido or OETH
+    /// This is not the ARM's withdrawal queue
+    function _externalWithdrawQueue() internal view virtual returns (uint256 assets);
+
+    /// @notice Calculates the amount of shares for a given amount of liquidity assets
+    function convertToShares(uint256 assets) public view returns (uint256 shares) {
+        uint256 totalAssetsMem = totalAssets();
+        shares = (totalAssetsMem == 0) ? assets : (assets * totalSupply()) / totalAssetsMem;
+    }
+
+    /// @notice Calculates the amount of liquidity assets for a given amount of shares
+    function convertToAssets(uint256 shares) public view returns (uint256 assets) {
+        assets = (shares * totalAssets()) / totalSupply();
+    }
+
+    /// @notice Set the Liquidity Provider Controller contract address.
+    /// Set to a zero address to disable the controller.
+    function setLiquidityProviderController(address _liquidityProviderController) external onlyOwner {
+        liquidityProviderController = _liquidityProviderController;
+
+        emit LiquidityProviderControllerUpdated(_liquidityProviderController);
+    }
+
+    ////////////////////////////////////////////////////
+    ///         Performance Fee Functions
+    ////////////////////////////////////////////////////
+
+    /// @notice Owner sets the performance fee on increased assets
+    /// @param _fee The performance fee measured in basis points (1/100th of a percent)
+    /// 10,000 = 100% performance fee
+    /// 500 = 5% performance fee
+    function setFee(uint256 _fee) external onlyOwner {
+        _setFee(_fee);
+    }
+
+    /// @notice Owner sets the account/contract that receives the performance fee
+    function setFeeCollector(address _feeCollector) external onlyOwner {
+        _setFeeCollector(_feeCollector);
+    }
+
+    function _setFee(uint256 _fee) internal {
+        require(_fee <= FEE_SCALE, "ARM: fee too high");
```

Comment:
🟠 Maybe `FEE_SCALE / 2` ? Don't want fees to be allowed to be set to 100%.

Reply:
I've made the max perf fee at 50%